File: src/core/geo/GeoDiscoveryEngine.ts
import { geoDiscoveryConfig } from "@/config/geoDiscovery";
import { GeoDiscoveryError, toGeoDiscoveryError } from "@/core/geo/geoError";
import { GeoLruCache } from "@/core/geo/lruCache";
import { normalizeGeoBatchQuery } from "@/core/geo/query";
import { GeoRequestRateGuard } from "@/core/geo/rateGuard";
import { withGeoRetry } from "@/core/geo/retry";
import type {
  GeoDiscoveryBatchQuery,
  GeoDiscoveryBatchResult,
  GeoDiscoveryNormalizedQuery,
  GeoDiscoveryPage,
  GeoDiscoveryQuery,
  GeoDiscoveryRepository,
  ServiceType
} from "@/types/geo";
import { geoTelemetry } from "@/utils/telemetry/geoTelemetry";

const stableCursorKey = (cursors: Partial<Record<ServiceType, string>>) =>
  Object.entries(cursors)
    .sort(([a], [b]) => a.localeCompare(b))
    .map(([serviceType, cursor]) => `${serviceType}:${cursor ?? "first"}`)
    .join(",");

const cacheKeyFor = (query: GeoDiscoveryNormalizedQuery) =>
  `${query.queryFingerprint}|cursor:${stableCursorKey(query.cursors) || "first"}`;

const markPage = <TEntity>(
  page: GeoDiscoveryPage<TEntity>,
  isStale: boolean,
  source: "cache",
): GeoDiscoveryPage<TEntity> => ({
  ...page,
  isStale,
  source
});

const markBatch = <TEntity>(
  result: GeoDiscoveryBatchResult<TEntity>,
  isStale: boolean,
): GeoDiscoveryBatchResult<TEntity> => ({
  ...result,
  results: Object.fromEntries(
    Object.entries(result.results).map(([serviceType, page]) => [
      serviceType,
      page ? markPage(page as GeoDiscoveryPage<TEntity>, isStale, "cache") : page
    ]),
  ),
  source: "cache"
});

export class GeoDiscoveryEngine {
  private cache = new GeoLruCache<GeoDiscoveryBatchResult<unknown>>();
  private inFlight = new Map<string, Promise<GeoDiscoveryBatchResult<unknown>>>();
  private rateGuard = new GeoRequestRateGuard();

  constructor(private readonly repository: GeoDiscoveryRepository) {}

  async getNearby<TEntity = unknown>(
    query: GeoDiscoveryQuery,
    signal?: AbortSignal,
  ): Promise<GeoDiscoveryPage<TEntity>> {
    const batchResult = await this.getNearbyBatch<TEntity>(
      {
        ...query,
        cursors: query.cursor ? { [query.serviceType]: query.cursor } : undefined,
        serviceTypes: [query.serviceType]
      },
      signal,
    );
    const page = batchResult.results[query.serviceType];

    if (!page) {
      throw new GeoDiscoveryError(`No ${query.serviceType} results returned`, {
        code: "unknown",
        retryable: true
      });
    }

    return page as GeoDiscoveryPage<TEntity>;
  }

  async getNearbyBatch<TEntity = unknown>(
    query: GeoDiscoveryBatchQuery,
    signal?: AbortSignal,
  ): Promise<GeoDiscoveryBatchResult<TEntity>> {
    const normalized = normalizeGeoBatchQuery(query);

    if (normalized.serviceTypes.length === 0) {
      throw new GeoDiscoveryError("At least one service type is required", {
        code: "unknown",
        retryable: false
      });
    }

    geoTelemetry.event(
      normalized.serviceTypes.length > 1 ? "geo_batch_search_requested" : "geo_search_requested",
      {
        geocell: normalized.roundedGeocell,
        pageSize: normalized.pageSize,
        queryFingerprint: normalized.queryFingerprint,
        radiusKm: normalized.radiusKm,
        schemaVersion: normalized.schemaVersion,
        serviceTypes: normalized.serviceTypes.join(",")
      },
    );

    return this.execute<TEntity>(normalized, signal);
  }

  clearCache() {
    this.cache.clear();
  }

  private async execute<TEntity>(
    normalized: GeoDiscoveryNormalizedQuery,
    signal?: AbortSignal,
  ): Promise<GeoDiscoveryBatchResult<TEntity>> {
    const key = cacheKeyFor(normalized);
    const cached = this.cache.get(key);

    if (cached.freshness === "fresh" && cached.value) {
      geoTelemetry.event("geo_cache_hit", {
        cacheHit: true,
        geocell: normalized.roundedGeocell,
        queryFingerprint: normalized.queryFingerprint,
        serviceTypes: normalized.serviceTypes.join(",")
      });
      return markBatch(cached.value as GeoDiscoveryBatchResult<TEntity>, false);
    }

    try {
      const result = await this.fetchNetwork<TEntity>(normalized, key, signal);
      this.cache.set(key, result as GeoDiscoveryBatchResult<unknown>);
      return result;
    } catch (error) {
      const geoError = toGeoDiscoveryError(error);

      if (geoError.code === "invalid_cursor") {
        geoTelemetry.warn("geo_cursor_invalidated", {
          code: geoError.code,
          geocell: normalized.roundedGeocell,
          queryFingerprint: normalized.queryFingerprint,
          serviceTypes: normalized.serviceTypes.join(",")
        });
        const firstPageQuery = { ...normalized, cursors: {} };
        const firstPageKey = cacheKeyFor(firstPageQuery);
        const result = await this.fetchNetwork<TEntity>(firstPageQuery, firstPageKey, signal);
        const cursorInvalidatedResult = {
          ...result,
          results: Object.fromEntries(
            Object.entries(result.results).map(([serviceType, page]) => [
              serviceType,
              page ? { ...page, cursorInvalidated: true } : page
            ]),
          )
        } as GeoDiscoveryBatchResult<TEntity>;
        this.cache.set(firstPageKey, cursorInvalidatedResult as GeoDiscoveryBatchResult<unknown>);
        return cursorInvalidatedResult;
      }

      if (cached.freshness === "stale" && cached.value) {
        geoTelemetry.warn("geo_cache_stale_served", {
          code: geoError.code,
          geocell: normalized.roundedGeocell,
          isStale: true,
          queryFingerprint: normalized.queryFingerprint,
          serviceTypes: normalized.serviceTypes.join(",")
        });
        return markBatch(cached.value as GeoDiscoveryBatchResult<TEntity>, true);
      }

      geoTelemetry.error("geo_search_failed", {
        code: geoError.code,
        geocell: normalized.roundedGeocell,
        queryFingerprint: normalized.queryFingerprint,
        serviceTypes: normalized.serviceTypes.join(","),
        status: "failed"
      });

      throw geoError;
    }
  }

  private async fetchNetwork<TEntity>(
    normalized: GeoDiscoveryNormalizedQuery,
    key: string,
    signal?: AbortSignal,
  ): Promise<GeoDiscoveryBatchResult<TEntity>> {
    const activeRequest = this.inFlight.get(key);

    if (activeRequest) {
      return activeRequest as Promise<GeoDiscoveryBatchResult<TEntity>>;
    }

    const startedAt = Date.now();
    const promise = (async () => {
      await this.rateGuard.waitForSlot(normalized.queryFingerprint, signal);

      const result = await withGeoRetry(
        () => this.repository.searchNearby<TEntity>(normalized, signal),
        {
          queryFingerprint: normalized.queryFingerprint,
          serviceTypes: normalized.serviceTypes.join(","),
          signal
        },
      );

      if (result.schemaVersion !== geoDiscoveryConfig.api.schemaVersion) {
        throw new GeoDiscoveryError("Geo discovery schema version is unsupported", {
          code: "schema_version_unsupported",
          retryable: false
        });
      }

      geoTelemetry.event("geo_search_succeeded", {
        durationMs: Date.now() - startedAt,
        emptyResult: Object.values(result.results).every((page) => !page || page.items.length === 0),
        geocell: normalized.roundedGeocell,
        queryFingerprint: normalized.queryFingerprint,
        schemaVersion: result.schemaVersion,
        serviceTypes: normalized.serviceTypes.join(","),
        source: result.source
      });

      return result;
    })();

    this.inFlight.set(key, promise as Promise<GeoDiscoveryBatchResult<unknown>>);

    try {
      return await promise;
    } finally {
      this.inFlight.delete(key);
    }
  }
}

File: lib/core/utils/geo_discovery/geo_discovery_engine.dart
import 'package:dio/dio.dart';

import '../../../config/app_config.dart';
import '../../../config/geo_discovery_config.dart';
import '../telemetry.dart';
import 'geo_cache.dart';
import 'geo_types.dart';
import 'query_normalizer.dart';
import 'rate_guard.dart';
import 'retry.dart';

class GeoDiscoveryEngine {
  GeoDiscoveryEngine({
    required GeoDiscoveryRepository repository,
    required GeoDiscoveryCache cache,
  }) : _repository = repository,
       _cache = cache;

  final GeoDiscoveryRepository _repository;
  final GeoDiscoveryCache _cache;
  final _inFlight =
      <String, Future<GeoDiscoveryBatchResult<Map<String, Object?>>>>{};
  final _lastFailedAtByFingerprint = <String, DateTime>{};
  final _lastSuccessfulByKey =
      <String, GeoDiscoveryBatchResult<Map<String, Object?>>>{};
  final _rateGuard = GeoRequestRateGuard();

  Future<GeoDiscoveryPage<Map<String, Object?>>> getNearby(
    GeoDiscoveryQuery query, {
    CancelToken? cancelToken,
  }) async {
    final result = await getNearbyBatch(
      GeoDiscoveryBatchQuery(
        cursors: query.cursor == null
            ? null
            : {query.serviceType: query.cursor!},
        filters: query.filters,
        latitude: query.latitude,
        longitude: query.longitude,
        pageSize: query.pageSize,
        radiusKm: query.radiusKm,
        requestId: query.requestId,
        serviceTypes: [query.serviceType],
        sort: query.sort,
      ),
      cancelToken: cancelToken,
    );
    final page = result.results[query.serviceType];
    if (page == null) {
      throw GeoDiscoveryError(
        'No ${query.serviceType.apiValue} results returned.',
        retryable: true,
      );
    }
    return page;
  }

  Future<GeoDiscoveryBatchResult<Map<String, Object?>>> getNearbyBatch(
    GeoDiscoveryBatchQuery query, {
    CancelToken? cancelToken,
  }) async {
    final normalized = normalizeGeoBatchQuery(query);
    if (normalized.serviceTypes.isEmpty) {
      throw const GeoDiscoveryError(
        'At least one service type is required.',
        retryable: false,
      );
    }

    telemetry.event(
      normalized.serviceTypes.length > 1
          ? TelemetryEvent.geoBatchSearchRequested
          : TelemetryEvent.geoSearchRequested,
      {
        'geocell': normalized.roundedGeocell,
        'pageSize': normalized.pageSize,
        'queryFingerprint': normalized.queryFingerprint,
        'radiusKm': normalized.radiusKm,
        'schemaVersion': normalized.schemaVersion,
        'serviceTypes': normalized.serviceTypes
            .map((entry) => entry.apiValue)
            .join(','),
        'sourceMode': AppConfig.geoRuntimeMode,
      },
    );

    return _execute(normalized, cancelToken: cancelToken);
  }

  Future<void> clearCache() => _cache.clear();

  Future<GeoDiscoveryBatchResult<Map<String, Object?>>> _execute(
    GeoDiscoveryNormalizedQuery normalized, {
    CancelToken? cancelToken,
  }) async {
    final key = _cacheKeyFor(normalized);
    final cached = _cache.get(key);

    if (cached.freshness == CacheFreshness.fresh && cached.value != null) {
      _lastFailedAtByFingerprint.remove(normalized.queryFingerprint);
      _lastSuccessfulByKey[key] = cached.value!;
      telemetry.event(TelemetryEvent.geoCacheHit, {
        'cacheHit': true,
        'geocell': normalized.roundedGeocell,
        'queryFingerprint': normalized.queryFingerprint,
        'serviceTypes': normalized.serviceTypes
            .map((entry) => entry.apiValue)
            .join(','),
      });
      return _markBatch(cached.value!, isStale: false);
    }

    final cooldownFallback = _cooldownFallbackFor(
      key,
      normalized,
      cached.value,
    );
    if (cooldownFallback != null) {
      return cooldownFallback;
    }

    try {
      final result = await _fetchNetwork(
        normalized,
        key,
        cancelToken: cancelToken,
      );
      await _cache.set(key, result);
      _lastFailedAtByFingerprint.remove(normalized.queryFingerprint);
      _lastSuccessfulByKey[key] = result;
      return result;
    } catch (error) {
      final geoError = toGeoDiscoveryError(error);
      if (geoError.code == GeoFailureCode.aborted) {
        final fallback = cached.value ?? _lastSuccessfulByKey[key];
        if (fallback != null) {
          return _markBatch(fallback, isStale: true);
        }
        throw geoError;
      }

      if (geoError.code == GeoFailureCode.invalidCursor) {
        telemetry.warn(TelemetryEvent.geoCursorInvalidated, {
          'code': geoError.code.apiValue,
          'geocell': normalized.roundedGeocell,
          'queryFingerprint': normalized.queryFingerprint,
        });
        final firstPage = GeoDiscoveryNormalizedQuery(
          cursors: const {},
          filtersByService: normalized.filtersByService,
          latitude: normalized.latitude,
          longitude: normalized.longitude,
          pageSize: normalized.pageSize,
          queryFingerprint: normalized.queryFingerprint,
          radiusKm: normalized.radiusKm,
          requestId: normalized.requestId,
          roundedGeocell: normalized.roundedGeocell,
          schemaVersion: normalized.schemaVersion,
          serviceTypes: normalized.serviceTypes,
          sort: normalized.sort,
        );
        final firstPageKey = _cacheKeyFor(firstPage);
        final result = await _fetchNetwork(
          firstPage,
          firstPageKey,
          cancelToken: cancelToken,
        );
        final marked = result.copyWith(
          results: result.results.map(
            (key, page) =>
                MapEntry(key, page.copyWith(cursorInvalidated: true)),
          ),
        );
        await _cache.set(firstPageKey, marked);
        return marked;
      }

      if (cached.freshness == CacheFreshness.stale && cached.value != null) {
        _lastFailedAtByFingerprint[normalized.queryFingerprint] =
            DateTime.now();
        telemetry.warn(TelemetryEvent.geoCacheStaleServed, {
          'code': geoError.code.apiValue,
          'geocell': normalized.roundedGeocell,
          'isStale': true,
          'queryFingerprint': normalized.queryFingerprint,
        });
        return _markBatch(cached.value!, isStale: true);
      }

      _lastFailedAtByFingerprint[normalized.queryFingerprint] = DateTime.now();
      if (geoError.code == GeoFailureCode.deploymentConfigError) {
        telemetry.error(TelemetryEvent.geoDeploymentMisconfiguration, {
          'code': geoError.code.apiValue,
          'geocell': normalized.roundedGeocell,
          'queryFingerprint': normalized.queryFingerprint,
        });
      }
      telemetry.error(TelemetryEvent.geoSearchFailed, {
        'code': geoError.code.apiValue,
        'geocell': normalized.roundedGeocell,
        'queryFingerprint': normalized.queryFingerprint,
        'status': 'failed',
      });
      throw geoError;
    }
  }

  GeoDiscoveryBatchResult<Map<String, Object?>>? _cooldownFallbackFor(
    String key,
    GeoDiscoveryNormalizedQuery normalized,
    GeoDiscoveryBatchResult<Map<String, Object?>>? cachedValue,
  ) {
    final failedAt = _lastFailedAtByFingerprint[normalized.queryFingerprint];
    if (failedAt == null) return null;

    final elapsed = DateTime.now().difference(failedAt);
    if (elapsed >= GeoDiscoveryConfig.failureCooldown) {
      _lastFailedAtByFingerprint.remove(normalized.queryFingerprint);
      return null;
    }

    final fallback = cachedValue ?? _lastSuccessfulByKey[key];
    telemetry.warn(TelemetryEvent.geoRetryCooldownTriggered, {
      'durationMs':
          (GeoDiscoveryConfig.failureCooldown - elapsed).inMilliseconds,
      'geocell': normalized.roundedGeocell,
      'queryFingerprint': normalized.queryFingerprint,
      'status': fallback == null ? 'blocked_without_cache' : 'served_cache',
    });

    if (fallback != null) {
      return _markBatch(fallback, isStale: true);
    }

    throw const GeoDiscoveryError(
      'Nearby discovery is cooling down after a failed request.',
      code: GeoFailureCode.networkError,
      retryable: false,
    );
  }

  Future<GeoDiscoveryBatchResult<Map<String, Object?>>> _fetchNetwork(
    GeoDiscoveryNormalizedQuery normalized,
    String key, {
    CancelToken? cancelToken,
  }) {
    final active = _inFlight[key];
    if (active != null) return active;

    final startedAt = DateTime.now();
    final promise = () async {
      await _rateGuard.waitForSlot(normalized.queryFingerprint);
      final result = await withGeoRetry(
        () => _repository.searchNearby(normalized, cancelToken: cancelToken),
        queryFingerprint: normalized.queryFingerprint,
        serviceTypes: normalized.serviceTypes
            .map((entry) => entry.apiValue)
            .join(','),
      );

      if (result.schemaVersion != GeoDiscoveryConfig.schemaVersion) {
        throw const GeoDiscoveryError(
          'Geo discovery schema version is unsupported.',
          code: GeoFailureCode.schemaVersionUnsupported,
          retryable: false,
        );
      }

      telemetry.event(TelemetryEvent.geoSearchSucceeded, {
        'durationMs': DateTime.now().difference(startedAt).inMilliseconds,
        'emptyResult': result.results.values.every(
          (page) => page.items.isEmpty,
        ),
        'geocell': normalized.roundedGeocell,
        'queryFingerprint': normalized.queryFingerprint,
        'schemaVersion': result.schemaVersion,
        'source': result.source.apiValue,
      });
      return result;
    }();

    _inFlight[key] = promise;
    return promise.whenComplete(() => _inFlight.remove(key));
  }

  String _cacheKeyFor(GeoDiscoveryNormalizedQuery query) {
    final cursors = query.cursors.entries.toList()
      ..sort((left, right) => left.key.apiValue.compareTo(right.key.apiValue));
    final cursorKey = cursors
        .map((entry) => '${entry.key.apiValue}:${entry.value}')
        .join(',');
    final cacheNamespace =
        'geo_v${GeoDiscoveryConfig.schemaVersion}|${AppConfig.geoRuntimeMode}|${AppConfig.apiBaseHost}';
    return '$cacheNamespace|${query.queryFingerprint}|cursor:${cursorKey.isEmpty ? 'first' : cursorKey}';
  }

  GeoDiscoveryBatchResult<Map<String, Object?>> _markBatch(
    GeoDiscoveryBatchResult<Map<String, Object?>> result, {
    required bool isStale,
  }) => result.copyWith(
    source: GeoResultSource.cache,
    results: result.results.map(
      (key, page) => MapEntry(
        key,
        page.copyWith(isStale: isStale, source: GeoResultSource.cache),
      ),
    ),
  );
}
